cmd/yggd/exec.go

```diff
@@ -5,135 +5,80 @@ import (
-		time.Sleep(delay)
-	}
-
 	stdout, err := cmd.StdoutPipe()
 	if err != nil {
-		log.Errorf("cannot connect to stdout: %v", err)
-		return
+		return fmt.Errorf("cannot connect to stdout: %v", err)
 	}
 
 	stderr, err := cmd.StderrPipe()
 	if err != nil {
-		log.Errorf("cannot connect to stderr: %v", err)
-		return
+		return fmt.Errorf("cannot connect to stderr: %v", err)
 	}
 
 	if err := cmd.Start(); err != nil {
-		log.Errorf("cannot start worker: %v: %v", file, err)
-		return
-	}
-	log.Debugf("started process: %v", cmd.Process.Pid)
+		return fmt.Errorf("cannot start worker: %v: %v", file, err)
 
-	go func() {
-		for {
-			buf := make([]byte, 4096)
-			n, err := stdout.Read(buf)
-			if n > 0 {
-				log.Tracef("[%v] %v", file, strings.TrimRight(string(buf), "\n\x00"))
-			}
-			if err != nil {
-				switch err {
-				case io.EOF:
-					log.Debugf("%v stdout reached EOF: %v", file, err)
-					return
-				default:
-					log.Errorf("cannot read from stdout: %v", err)
-					continue
-				}
-			}
-		}
-	}()
-
-	go func() {
-		for {
-			buf := make([]byte, 4096)
-			n, err := stderr.Read(buf)
-			if n > 0 {
-				log.Errorf("[%v] %v", file, strings.TrimRight(string(buf), "\n\x00"))
-			}
-			if err != nil {
-				switch err {
-				case io.EOF:
-					log.Debugf("%v stderr reached EOF: %v", file, err)
-					return
-				default:
-					log.Errorf("cannot read from stderr: %v", err)
-					continue
-				}
-			}
-		}
-	}()
-
-	pidDirPath := filepath.Join(yggdrasil.LocalstateDir, "run", yggdrasil.LongName, "workers")
-
-	if err := os.MkdirAll(pidDirPath, 0755); err != nil {
-		log.Errorf("cannot create directory: %v", err)
-		return
 	}
+	log.Debugf("started process: %v", cmd.Process.Pid)
 
-	if err := os.WriteFile(filepath.Join(pidDirPath, filepath.Base(file)+".pid"), []byte(fmt.Sprintf("%v", cmd.Process.Pid)), 0644); err != nil {
-		log.Errorf("cannot write to file: %v", err)
-		return
+	if started != nil {
+		go started(cmd.Process.Pid, stdout, stderr)
 	}
 
-	go watchProcess(cmd, delay, died)
+	return nil
 }
 
-func watchProcess(cmd *exec.Cmd, delay time.Duration, died chan int) {
-	log.Debugf("watching process: %v", cmd.Process.Pid)
-
-	state, err := cmd.Process.Wait()
+// waitProcess finds a process with the given pid and waits for it to exit.
+// If died is not nil, it is invoked on a goroutine when the process exits.
+func waitProcess(pid int, died func(pid int, state *os.ProcessState)) error {
+	process, err := os.FindProcess(pid)
 	if err != nil {
-		log.Errorf("process %v exited with error: %v", cmd.Process.Pid, err)
+		return fmt.Errorf("cannot find process with pid: %v", err)
 	}
 
-	died <- state.Pid()
-
-	if state.SystemTime() < time.Duration(1*time.Second) {
-		delay += 5 * time.Second
+	log.Debugf("waiting for process: %v", process.Pid)
+	state, err := process.Wait()
+	if err != nil {
+		return fmt.Errorf("process %v exited with error: %v", process.Pid, err)
 	}
-	if delay >= time.Duration(30*time.Second) {
-		delay = -1
+
+	if died != nil {
```

Comment:
Leading from my findings in the thread about `started`, I think I'd prefer to avoid the `Callback` suffix, but I agree that "died" is not the right name; "exited" or "stopped" is probably a better name.